### ISSUE TITLE
Missing dependency when installing iPython on Python 2.7

### ISSUE THREAD

**Issue by charettes** [sdispater/poetry#151]
Thanks for this project, I encountered the following missing `enum` dependency when installing iPython on Python 2.7.

Here's how I reproduced against poetry `0.10.0`.

```toml
[tool.poetry]
name = "test"
version = "0.10.0"
description = ""
authors = []

[tool.poetry.dependencies]
python = "2.7"
ipython = "=5.6.0"
```

And once `poetry install` is done I tried both `poetry run ipython` and `ipython` without success.

```
Traceback (most recent call last):
  File "/home/simon/.virtualenvs/tmp-33607c6fbabd8b9/bin/ipython", line 7, in <module>
    from IPython import start_ipython
  File "/home/simon/.virtualenvs/tmp-33607c6fbabd8b9/local/lib/python2.7/site-packages/IPython/__init__.py", line 48, in <module>
    from .core.application import Application
  File "/home/simon/.virtualenvs/tmp-33607c6fbabd8b9/local/lib/python2.7/site-packages/IPython/core/application.py", line 23, in <module>
    from traitlets.config.application import Application, catch_config_error
  File "/home/simon/.virtualenvs/tmp-33607c6fbabd8b9/local/lib/python2.7/site-packages/traitlets/__init__.py", line 1, in <module>
    from .traitlets import *
  File "/home/simon/.virtualenvs/tmp-33607c6fbabd8b9/local/lib/python2.7/site-packages/traitlets/traitlets.py", line 49, in <module>
    import enum
ImportError: No module named enum
```

I assume this is caused by `traitlets`'s usage of `extra_requires` for Python 2.

https://github.com/ipython/traitlets/blob/ee1a829fafad654601dc851dc1d4ec79da090913/setup.py#L92

**Comment by charettes** [comment]
By the way, I'd be happy to submit a fix for it if you could point me in the right direction.

**Comment by sdispater** [comment]
The dependencies are properly picked up on my end:

```text
poetry debug:resolve ipython=5.6.0

Resolution results:

  - appnope (0.1.0)
  - backports.shutil-get-terminal-size (1.0.0)
  - colorama (0.3.9)
  - decorator (4.3.0)
  - enum34 (1.1.6)
  - ipython-genutils (0.2.0)
  - pathlib2 (2.3.2)
  - pexpect (4.6.0)
  - pickleshare (0.7.4)
  - prompt-toolkit (1.0.15)
  - ptyprocess (0.5.2)
  - pygments (2.2.0)
  - scandir (1.7)
  - simplegeneric (0.8.1)
  - six (1.11.0)
  - traitlets (4.3.2)
  - wcwidth (0.1.7)
  - win-unicode-console (0.5)
  - ipython (5.6.0)
```

And `enum34` is picked up in the dependencies of `traitlets`:

```text
   1: fact: traitlets (4.3.2) depends on decorator (*)
   1: fact: traitlets (4.3.2) depends on ipython-genutils (*)
   1: fact: traitlets (4.3.2) depends on six (*)
   1: fact: traitlets (4.3.2) depends on enum34 (*)
```

**Comment by charettes** [comment]
FWIW it's reproducible by directly `add`'ing `traitlets`

```
simon@simon-laptop:~/.virtualenvs$ mktmpenv 
New python executable in /home/simon/.virtualenvs/tmp-c0fff0f469d0bf49/bin/python
Installing setuptools, pip, wheel...done.
virtualenvwrapper.user_scripts creating /home/simon/.virtualenvs/tmp-c0fff0f469d0bf49/bin/predeactivate
virtualenvwrapper.user_scripts creating /home/simon/.virtualenvs/tmp-c0fff0f469d0bf49/bin/postdeactivate
virtualenvwrapper.user_scripts creating /home/simon/.virtualenvs/tmp-c0fff0f469d0bf49/bin/preactivate
virtualenvwrapper.user_scripts creating /home/simon/.virtualenvs/tmp-c0fff0f469d0bf49/bin/postactivate
virtualenvwrapper.user_scripts creating /home/simon/.virtualenvs/tmp-c0fff0f469d0bf49/bin/get_env_details
This is a temporary environment. It will be deleted when you run 'deactivate'.
(tmp-c0fff0f469d0bf49) simon@simon-laptop:~/.virtualenvs/tmp-c0fff0f469d0bf49$ cp /tmp/pyproject.toml .
(tmp-c0fff0f469d0bf49) simon@simon-laptop:~/.virtualenvs/tmp-c0fff0f469d0bf49$ cat /tmp/pyproject.toml 
[tool.poetry]
name = "test"
version = "0.10.0"
description = ""
authors = []

[tool.poetry.dependencies]
python = "2.7"
(tmp-c0fff0f469d0bf49) simon@simon-laptop:~/.virtualenvs/tmp-c0fff0f469d0bf49$ poetry add traitlets==4.3.2

Updating dependencies
Resolving dependencies... (0.3s)


Package operations: 4 installs, 0 updates, 0 removals

Writing lock file

  - Installing decorator (4.3.0)
  - Installing ipython-genutils (0.2.0)
  - Installing six (1.11.0)
  - Installing traitlets (4.3.2)
(tmp-c0fff0f469d0bf49) simon@simon-laptop:~/.virtualenvs/tmp-c0fff0f469d0bf49$ python
Python 2.7.12 (default, Dec  4 2017, 14:50:18) 
[GCC 5.4.0 20160609] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import enum
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: No module named enum
>>> 
```

**Comment by sdispater** [comment]
I think I know where the problem is coming from.

`traitlets` defines `enum34` as a dependency twice:

```text
enum34; python_version=="2.7"
enum34; python_version=="3.3"
```

Currently, `poetry` assumes that every dependency appear only once so in this case it picks up the last one: `enum34; python_version=="3.3"`

That's why `enum34` is not installed for Python 2.7.

I am not sure about the best approach here. While I think the proper way to declare the `enum34` dependency should be `enum34; python_version in "2.7 3.3"`, `poetry` should be able to handle this.

**Comment by bertjwregeer** [comment]
@sdispater This is something that is likely to come up more often, for example the following is something that I have in a project for work:

```
pandas (~=0.19.0); (python_version==\"3.4\")
pandas; (python_version==\"2.7\" or python_version>=\"3.5\")
```

Due to the requirement to have a different panda's version for different Python versions.

So `panda`'s is defined twice, but with different versions pinned depending on the Python version.

**Comment by sdispater** [comment]
@bertjwregeer While

```
enum34; python_version=="2.7"
enum34; python_version=="3.3"
```

is easy to fix (I already a working fix which merges the markers),

```
pandas (~=0.19.0); (python_version==\"3.4\")
pandas; (python_version==\"2.7\" or python_version>=\"3.5\")
```

is another problem entirely since the version constraints for `pandas` are different, which can lead to a completely different dependency graph for different Python versions.

Right now, the dependency resolver assumes only one version for a given package, so it will not be an easy fix and will likely require tweaking the resolver.

**Comment by bertjwregeer** [comment]
@sdispater Yeah, that is why I did want to bring it up so that it could at least be documented as a possibility.

**Comment by sdispater** [comment]
The initial issue should be fixed in the latest release (`0.10.3`), so I am closing this one.

**Comment by charettes** [comment]
Awesome, thanks @sdispater!

**Comment by bertjwregeer** [comment]
@sdispater would you like me to open a new issue for that pandas example?

**Comment by sdispater** [comment]
@bertjwregeer I think #164 is the same issue so we can use it to track this
